### DISCUSSION TITLE
Support golang

### DISCUSSION THREAD

**Discussion by ITler** [projen/projen#2280] [General]
Hello,
I am new to projen.
While I have seen [projen Go bindings](https://github.com/projen/projen-go), I am wondering how to write my own (i.e.) Project definitions in Go and let `projen new my-go-based-project` create my template based on an [External module](https://github.com/projen/projen#projects-in-external-modules) on it.
I would like to keep out Javascript/Typescript from my code base as much as possible.
Do you have any example implementation/tutorial/docs available for that?

**Comment by mrgrain** [comment]
`projen new my-go-based-project` in Go is not something you would be able to do.
Projen projects need to be written in jsii. This is by design.

You can however write your abstractions in Go and include them like regular modules. But the flow would be something like

- `projen new a-jsii-project` 
- add `my-go-module` to dependencies
- use `my-go-module`

**Reply by ITler** [reply]
Hey. Thanks for answering. 

Unfortunately, I don't fully get it, yet.
I understood that if I want to provide a new project-type (`foo`) to be used like `projen new --from projen-foo foo` , I would need to create my project-type `foo` as a JSII project, [manually](https://aws.github.io/jsii/user-guides/lib-author/quick-start/set-up/) or using `projen new jsii` itself. 
Having the JSII project ready, would allow me at the end to publish a NPM module `projen-foo` to be consumed by `projen new --from projen-foo foo`. 
However the implementation for my `foo` 's project type (pipeline code, config files, whatever) itself has to be done with Typescript?! But where do the [projen-go bindings](https://github.com/projen/projen-go/tree/main/projen) come into play? I would want to use those to model the project type structure.

What do I miss?

**Reply by mrgrain** [reply]
The projen go bindings come into place when you want to write your projen config in go, see: https://github.com/projen/projen#do-i-have-to-write-my-configuration-in-javascript

We don't have a good example for this in Go at the moment (nor any other language, to be honest). The easiest way to "experience" this is at the moment probably `projen new python`. You can see how the config file is called `.projenrc.py` and written in Python. Imagine the same in Go.

PS: Would love to have a Go project in projen by default!
PPS: This is all quite complex and not well documented. Your reaction is perfectly normal.

**Reply by ITler** [reply]
Ha! Now I got it.
I don't need a `.projenrc.go` or similar. I just need to create a usual go project and use the go-bindings managed via `go.mod` ... No need to deal with `npx projen` (as it would try to use Node to compile the go code) just `go run .`. And with the code listed below, a plain project with a file called `foo.txt` is created
```
package main

import (
	"github.com/aws/jsii-runtime-go"
	"github.com/projen/projen-go/projen"
)

func main() {
	prj := projen.NewProject(&projen.ProjectOptions{
		Name:   jsii.String("foo"),
		Outdir: jsii.String("synth"),
	})

	projen.NewTextFile(prj, jsii.String("foo.txt"), &projen.TextFileOptions{
		Lines: &[]*string{jsii.String("line1")},
	})
	prj.Synth()
}
```
From here I could go with regular go modules to model my project structure as I would need it.

Is this the way it suppose to work from your perspective?. Would you have more insights to share?

**Reply by mrgrain** [reply]
Yes, that's the way to go.

One limitation will be the task runner, which requires the projen binary to do commands like `projen test`.

Having a jsii module comes with these 2 advantages. If you're not interested in them, there's little reason to do it:
- support for `projen new --from`
- reusability of projects across languagues

**Reply by ITler** [reply]
Ok. Sounds good. I could image, that coding my project setup will also lead to creation of a custom go-based CLI tool (to replace `npx projen` command), that would nicely wrap calls to my custom project setup and to create the `projenrc.go` in the bootstrapped project for further customizations. 

Yeah, for my actual use case, the JSII advantages you were mentioning are not really relevant.

However I am not sure what you mean with TaskRunner? I thought tasks at the end of the day are reflected in `scripts` section in `package.json`? Or what does `projen test` give me as an advantage?

**Reply by mrgrain** [reply]
> However I am not sure what you mean with TaskRunner? I thought tasks at the end of the day are reflected in scripts section in package.json? Or what does projen test give me as an advantage?

Projen comes with a built-in task runner. So when you do `new Task('my-task-name')` (or whatever the Go equivalent is), it can be called with `projen my-task-name`. The `script` section of `package.json` is only a bonus added to Node.js projects.

> Ok. Sounds good. I could image, that coding my project setup will also lead to creation of a custom go-based CLI tool (to replace npx projen command), that would nicely wrap calls to my custom project setup and to create the projenrc.go in the bootstrapped project for further customizations.

I wonder what needs to be done to publish projen as a go executable. We already do that for python... 🤔

**Reply by ITler** [reply]
I see.. Currently I think of totally replacing `npx projen` with my own CLI tool, as from what I see today, I personally wouldn't need all the capabilities that `npx projen` brings - I don't like that, but at least as long as there is no built-in support for go. 
I am not a go-pro, but if projen would aim for supporting go, the go variant of the projen binary would probably change the 'user interface'

```
# ...from
npx projen new --from some-projen-module@latest your_project_from_module

# to something like
go install github.com/projen/projen-go/projen-cli@latest # install projen
go install github.com/owner/some-projen-go-module@latest # install a custom module
projen-go new your_project_from_module # create custom project
```

However, I couldn't figure out, yet, what is done for python. I **guess** the python-projen executable is created when projen-jsii-project is published/released (not sure about the terms and jsii project lifecycle). I got stuck when looking into the repos and some files referring to some `../lib/<...>` sources that seem to not be included to the repo here
